File: tests/Scribe.Core.Tests/DictionaryTermVariantsTests.cs
using Scribe.Core.PostProcessing;
using Xunit;

namespace Scribe.Core.Tests;

/// <summary>
/// Guards the boundary that keeps auto-learning honest: a generated pattern must be something the
/// recognizer was actually observed to produce, and must never be able to rewrite ordinary prose.
/// </summary>
public sealed class DictionaryTermVariantsTests
{
    [Theory]
    [InlineData("CSU", "c s u")]
    [InlineData("ATU", "a t u")]
    [InlineData("MCAP", "m c a p")]
    [InlineData(".NET", "n e t")]
    public void Acronyms_are_spelled_out(string term, string expected) =>
        Assert.Contains(expected, DictionaryTermVariants.For(term));

    [Theory]
    [InlineData("AI")]
    [InlineData("IQ")]
    [InlineData("PR")]
    public void Two_letter_acronyms_are_skipped(string term) =>
        // Their spelled form is two single letters, which collides with "a", "I" and friends.
        Assert.Empty(DictionaryTermVariants.For(term));

    [Theory]
    [InlineData("WebIQ", "web iq")]
    [InlineData("GitHub", "git hub")]
    [InlineData("JavaScript", "java script")]
    [InlineData("DeepSeek", "deep seek")]
    public void Compounds_are_split(string term, string expected) =>
        Assert.Contains(expected, DictionaryTermVariants.For(term));

    [Theory]
    [InlineData("AndThen")]     // "and then" is prose, not a rendering fix
    [InlineData("TheOther")]
    [InlineData("IfNot")]
    [InlineData("LinkedIn")]    // "the issue linked in the PR" must survive untouched
    public void Compounds_made_of_everyday_words_are_rejected(string term) =>
        Assert.Empty(DictionaryTermVariants.For(term));

    [Fact]
    public void A_generated_pattern_never_equals_its_replacement()
    {
        string[] terms = ["CSU", "WebIQ", "GitHub", "MCAP", "JavaScript", "ReBAC"];

        foreach (var term in terms)
        {
            Assert.DoesNotContain(
                DictionaryTermVariants.For(term),
                pattern => string.Equals(pattern, term, StringComparison.OrdinalIgnoreCase));
        }
    }

    [Fact]
    public void A_generated_pattern_differs_from_the_term_only_in_case_and_spacing()
    {
        string[] terms = ["CSU", "ATU", "WebIQ", "GitHub", "JavaScript", "MCAP"];

        foreach (var term in terms)
        {
            var squashedTerm = Squash(term);
            foreach (var pattern in DictionaryTermVariants.For(term))
            {
                // This is what makes a generated rule safe: it can only ever restore rendering,
                // never substitute a different word.
                Assert.Equal(squashedTerm, Squash(pattern));
            }
        }
    }

    [Theory]
    [InlineData("hello")]
    [InlineData("Hello")]
    [InlineData("42")]
    [InlineData("")]
    [InlineData("   ")]
    public void Ordinary_tokens_yield_nothing(string term) =>
        Assert.Empty(DictionaryTermVariants.For(term));

    private static string Squash(string value) =>
        new(value.Where(char.IsLetterOrDigit).Select(char.ToLowerInvariant).ToArray());
}

File: macos/Scribe/Tests/ScribeTests/DictionaryTermVariantsTests.swift
import XCTest
@testable import Scribe

/// Guards the boundary that keeps auto-learning honest: a generated pattern must be something the
/// recognizer was actually observed to produce, and must never be able to rewrite ordinary prose.
final class DictionaryTermVariantsTests: XCTestCase {
    func testAcronymsAreSpelledOut() {
        let cases: [(String, String)] = [
            ("CSU", "c s u"),
            ("ATU", "a t u"),
            ("MCAP", "m c a p"),
            (".NET", "n e t"),
        ]
        for (term, expected) in cases {
            XCTAssertTrue(DictionaryTermVariants.variants(for: term).contains(expected), term)
        }
    }

    func testTwoLetterAcronymsAreSkipped() {
        // Their spelled form is two single letters, which collides with "a", "I" and friends.
        for term in ["AI", "IQ", "PR"] {
            XCTAssertTrue(DictionaryTermVariants.variants(for: term).isEmpty, term)
        }
    }

    func testCompoundsAreSplit() {
        let cases: [(String, String)] = [
            ("WebIQ", "web iq"),
            ("GitHub", "git hub"),
            ("JavaScript", "java script"),
            ("DeepSeek", "deep seek"),
        ]
        for (term, expected) in cases {
            XCTAssertTrue(DictionaryTermVariants.variants(for: term).contains(expected), term)
        }
    }

    func testCompoundsMadeOfEverydayWordsAreRejected() {
        // "and then" is prose, not a rendering fix; "the issue linked in the PR" must survive untouched.
        for term in ["AndThen", "TheOther", "IfNot", "LinkedIn"] {
            XCTAssertTrue(DictionaryTermVariants.variants(for: term).isEmpty, term)
        }
    }

    func testAGeneratedPatternNeverEqualsItsReplacement() {
        let terms = ["CSU", "WebIQ", "GitHub", "MCAP", "JavaScript", "ReBAC"]

        for term in terms {
            XCTAssertFalse(
                DictionaryTermVariants.variants(for: term).contains {
                    $0.caseInsensitiveCompare(term) == .orderedSame
                },
                term)
        }
    }

    func testAGeneratedPatternDiffersFromTheTermOnlyInCaseAndSpacing() {
        let terms = ["CSU", "ATU", "WebIQ", "GitHub", "JavaScript", "MCAP"]

        for term in terms {
            let squashedTerm = squash(term)
            for pattern in DictionaryTermVariants.variants(for: term) {
                // This is what makes a generated rule safe: it can only ever restore rendering,
                // never substitute a different word.
                XCTAssertEqual(squash(pattern), squashedTerm, term)
            }
        }
    }

    func testOrdinaryTokensYieldNothing() {
        for term in ["hello", "Hello", "42", "", "   "] {
            XCTAssertTrue(DictionaryTermVariants.variants(for: term).isEmpty, "'\(term)'")
        }
    }

    private func squash(_ value: String) -> String {
        String(value.filter { $0.isLetter || $0.isNumber }).lowercased()
    }
}
